src/Chapter3.hs

```diff
@@ -480,6 +519,36 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+newtype Castle = MkCastle
+    { castleName :: String
+    }
+
+data Building = Church | Library
+
+newtype House = MkHouse
+    { housePeople :: Int
+    }
+
+data City = MkCity
+    { cityCastle   :: Maybe Castle
+    , cityWalls    :: Int
+    , cityBuilding :: Building
+    , cityHouses   :: [House]
+    }
+
+buildCastle :: City -> Castle -> City
+buildCastle city castle = city { cityCastle = Just castle }
+
+buildHouse :: City -> House -> City
+buildHouse city house = city { cityHouses = house : cityHouses city }
+
+buildWalls :: City -> City
+buildWalls city =
+    let peopleCount = sum (map housePeople (cityHouses city))
+    in if peopleCount >= 10
+        then city { cityWalls = cityWalls city + 1 }
```

Comment:
I see that you don't check is there are any castles, so this leaks the unrepresentable state 🙂